include/zephyr/net/coap_service.h

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2023 Basalte bv
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief CoAP Service API
+ *
+ * An API for applications to respond to CoAP requests
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_COAP_SERVICE_H_
+#define ZEPHYR_INCLUDE_NET_COAP_SERVICE_H_
+
+#include <zephyr/net/coap.h>
+#include <zephyr/sys/iterable_sections.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief CoAP Service API
+ * @defgroup coap_service CoAP service API
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * @name CoAP Service configuration flags
+ * @anchor COAP_SERVICE_FLAGS
+ * @{
+ */
+
+/** Start the service on boot. */
+#define COAP_SERVICE_AUTOSTART		BIT(0)
+
+/** @} */
+
+/** @cond INTERNAL_HIDDEN */
+
+struct coap_service_data {
+	int sock_fd;
+	struct coap_observer observers[CONFIG_COAP_SERVICE_OBSERVERS];
+	struct coap_pending pending[CONFIG_COAP_SERVICE_PENDING_MESSAGES];
+};
+
+struct coap_service {
+	const char *name;
+	const char *host;
+	uint16_t *port;
+	uint8_t flags;
+	struct coap_resource *res_begin;
+	struct coap_resource *res_end;
+	struct coap_service_data *data;
+};
+
+#define __z_coap_service_define(_name, _host, _port, _flags, _res_begin, _res_end)		\
+	static struct coap_service_data coap_service_data_##_name;				\
+	const STRUCT_SECTION_ITERABLE(coap_service, _name) = {					\
+		.name = STRINGIFY(_name),							\
+		.host = _host,									\
+		.port = (uint16_t *)(_port),							\
+		.flags = _flags,								\
+		.res_begin = (_res_begin),							\
+		.res_end = (_res_end),								\
+		.data = &coap_service_data_##_name,						\
+	}
+
+/** @endcond */
+
+/**
+ * @brief Define a static CoAP resource owned by the service named @p _service .
+ *
+ * @param _name Name of the resource.
+ * @param _service Name of the associated service.
+ */
+#define COAP_RESOURCE_DEFINE(_name, _service, ...)						\
+	STRUCT_SECTION_ITERABLE_ALTERNATE(coap_resource_##_service, coap_resource, _name)	\
+		= __VA_ARGS__
+
+/**
+ * @brief Define a CoAP service with static resources.
+ *
+ * @note The @p _host parameter can be `NULL`. If not, it is used to specify an IP address either in
+ * IPv4 or IPv6 format a fully-qualified hostname or a virtual host, otherwise the any address is
+ * used.
+ *
+ * @note The @p _port parameter must be non-`NULL`. It points to a location that specifies the port
+ * number to use for the service. If the specified port number is zero, then an ephemeral port
+ * number will be used and the actual port number assigned will be written back to memory. For
+ * ephemeral port numbers, the memory pointed to by @p _port must be writeable.
+ *
+ * @param _name Name of the service.
+ * @param _host IP address or hostname associated with the service.
+ * @param[inout] _port Pointer to port associated with the service.
+ * @param _flags Configuration flags @see @ref COAP_SERVICE_FLAGS.
+ */
+#define COAP_SERVICE_DEFINE(_name, _host, _port, _flags)					\
```

Comment:
Does this mean that all the services need to be defined at built time?

Reply:
That's the idea, yes. They can be started/stopped at runtime, but a single server thread is used for all services.

Reply:
Both services and resources are placed in predefined linker sections.

Reply:
Right, I'll add a note to the function brief, and add a check/assert.